planner/core/hints.go

```diff
@@ -154,3 +165,121 @@ func (p *BlockHintProcessor) getCurrentStmtHints(hints []*ast.TableOptimizerHint
 	}
 	return p.QbHints[currentOffset]
 }
+
+func restoreOptimizerHint(hint *ast.TableOptimizerHint) string {
+	var sb strings.Builder
+	ctx := format.NewRestoreCtx(format.DefaultRestoreFlags, &sb)
+	err := hint.Restore(ctx)
+	// There won't be any error for optimizer hint.
+	if err != nil {
+		logutil.BgLogger().Warn("restore hint failed", zap.Error(err))
+	}
+	return sb.String()
+}
+
+// GenHintsFromPhysicalPlan generates hints from physical plan.
+func GenHintsFromPhysicalPlan(p Plan) string {
+	var hints []*ast.TableOptimizerHint
+	switch pp := p.(type) {
+	case *Update:
+		hints = genHintsFromPhysicalPlan(pp.SelectPlan, typeUpdate)
+	case *Delete:
+		hints = genHintsFromPhysicalPlan(pp.SelectPlan, typeDelete)
+	case PhysicalPlan:
+		hints = genHintsFromPhysicalPlan(pp, typeSelect)
+	}
+	hintsStr := make([]string, 0, len(hints))
+	for _, hint := range hints {
+		hintsStr = append(hintsStr, restoreOptimizerHint(hint))
+	}
+	return strings.Join(hintsStr, ", ")
+}
+
+func generateQBName(nodeType nodeType, offset int) model.CIStr {
```

Comment:
It's better to keep the same variable name to improve the code readability.
```suggestion
func generateQBName(nodeType nodeType, blockOffset int) model.CIStr {
```